src/code42cli/output_formats.py

```diff
@@ -85,57 +69,213 @@ def _requires_list_output(self):
-                "index": True,
-                "force_ascii": False,
-                "default_handler": str,
-            }
-            defaults.update(kwargs)
-            return df.to_json(**defaults)
+    def iter_rows(self, dfs, columns=None):
+        """
+        Accepts a pandas DataFrame or list/generator of DataFrames and yields each
+        'row' of the DataFrame as a dict, calling the `checkpoint_func` on each row
+        after it has been yielded.
+
+        Accepts an optional list of column names that filter
+        columns in the yielded results.
+        """
+        dfs = self._ensure_iterable(dfs)
+        for df in dfs:
+            # convert pandas' default null (numpy.NaN) to None
+            df = df.astype(object).where(notnull, None)
+            if columns:
+                filtered = self._select_columns(df, columns)
+            else:
+                filtered = df
+            for full_event, filtered_event in zip(
+                df.to_dict("records"), filtered.to_dict("records")
+            ):
+                yield filtered_event
+                self.checkpoint_func(full_event)
+
+    def get_formatted_output(self, dfs, columns=None, **kwargs):
+        """
+        Accepts a pandas DataFrame or list/generator of DataFrames and formats and yields
+        the results line by line to the caller as a generator.
+
+        Accepts an optional list of column names that filter columns in the yielded
+        results.
+
+        Any additional kwargs provided will be passed to the underlying format method
+        if customizations are required.
+        """
+        if self.output_format == OutputFormat.TABLE:
+            yield from self._iter_table(dfs, columns=columns, **kwargs)
 
         elif self.output_format == OutputFormat.CSV:
-            defaults = {"index": False, "line_terminator": "\n"}
-            defaults.update(kwargs)
-            df = df.fillna("")
-            return df.to_csv(**defaults)
+            yield from self._iter_csv(dfs, columns=columns, **kwargs)
 
-        elif self.output_format == OutputFormat.TABLE:
-            defaults = {"index": False}
-            defaults.update(kwargs)
-            df = df.fillna("")
-            return df.to_string(**defaults)
+        elif self.output_format == OutputFormat.JSON:
+            kwargs = {"indent": 4, **kwargs}
+            yield from self._iter_json(dfs, columns=columns, **kwargs)
+
+        elif self.output_format == OutputFormat.RAW:
+            yield from self._iter_json(dfs, columns=columns, **kwargs)
 
         else:
-            raise ValueError(
+            raise Code42CLIError(
                 f"DataFrameOutputFormatter received an invalid format: {self.output_format}"
             )
 
-    def echo_formatted_dataframe(self, df, **kwargs):
-        str_output = self.get_formatted_output(df, **kwargs)
-        if len(df) <= OUTPUT_VIA_PAGER_THRESHOLD:
-            click.echo(str_output)
+    def echo_formatted_dataframes(
```

Comment:
If no results are found, this prints nothing.  Not sure if we want to echo the usual "No results found." here or in the method where this `echo_formatted_dataframes()` gets called.